packages/lodestar/src/sync/regular/oneRangeAhead/fetcher.ts

```diff
@@ -0,0 +1,106 @@
+  private rangeEnd: Slot = 0;
+  private getPeers: () => Promise<PeerId[]>;
+
+  constructor(options: Partial<IRegularSyncOptions>, modules: IRegularSyncModules, getPeers: () => Promise<PeerId[]>) {
+    this.config = modules.config;
+    this.network = modules.network;
+    this.chain = modules.chain;
+    this.logger = modules.logger;
+    this.opts = deepmerge(defaultOptions, options);
+    this.getPeers = getPeers;
+    this.lastFetchCheckpoint = {blockRoot: ZERO_HASH, slot: 0};
+  }
+
+  public setLastProcessedBlock(lastProcessedBlock: ISyncCheckpoint): void {
+    this.lastFetchCheckpoint = lastProcessedBlock;
+  }
+
+  /**
+   * Get next block range.
+   */
+  public async next(): Promise<SignedBeaconBlock[]> {
+    this.updateNextRange();
+    let result: SignedBeaconBlock[] | null = null;
+    while (!result || !result!.length) {
+      let slotRange: ISlotRange | null = null;
+      try {
+        const peers = await this.getPeers();
+        if (result && !result.length) await this.handleEmptyRange(peers);
+        slotRange = {start: this.rangeStart, end: this.rangeEnd};
+        result = await getBlockRange(this.logger, this.network.reqResp, peers, slotRange);
+      } catch (e) {
+        this.logger.debug("Regular Sync: Failed to get block range ", {...(slotRange ?? {}), error: e.message});
+        // sync is stopped for whatever reasons
+        if (e instanceof ErrorAborted) return [];
+        result = null;
+      }
+    }
+    // success
+    const lastBlock = result[result.length - 1].message;
+    this.lastFetchCheckpoint = {blockRoot: this.config.types.BeaconBlock.hashTreeRoot(lastBlock), slot: lastBlock.slot};
+    return result!;
+  }
+
+  // always set range based on last fetch block bc sometimes the previous fetch may not return all blocks
+  private updateNextRange(): void {
+    this.rangeStart = this.lastFetchCheckpoint.slot + 1;
+    this.rangeEnd = this.rangeStart;
+    this.rangeEnd = this.getNewTarget();
+  }
+
+  private async handleEmptyRange(peers: PeerId[] = []): Promise<void> {
+    if (!this.getPeers.length) {
+      return;
+    }
+    const range = {start: this.rangeStart, end: this.rangeEnd};
+    const peerHeadSlot = Math.max(...peers.map((peer) => this.network.peerMetadata.getStatus(peer)?.headSlot ?? 0));
+    this.logger.verbose("Regular Sync: Not found any blocks for range", {range});
+    if (range.end <= peerHeadSlot) {
+      // range contains skipped slots, query for next range
+      this.logger.verbose("Regular Sync: queried range is behind peer head, fetch next range", {
+        ...range,
+        peerHead: peerHeadSlot,
+      });
+      // don't trust empty range as it's rarely happen, peer may return it incorrectly or not up to date
+      // same range start, expand range end
+      this.rangeEnd = this.getNewTarget();
+    } else {
+      this.logger.verbose("Regular Sync: Queried range passed peer head, sleep then try again", {
+        range: JSON.stringify(range),
```

Comment:
`range` is stringified here but not above. Unless there's a strong reason to stringify it, pass the entire JSON to the logger